mlflow/store/tracking/rest_store.py

```diff
@@ -289,8 +290,9 @@ def get_experiment_by_name(self, experiment_name):
                 databricks_pb2.RESOURCE_DOES_NOT_EXIST
             ):
                 return None
-            elif e.error_code == databricks_pb2.ErrorCode.Name(
-                databricks_pb2.REQUEST_LIMIT_EXCEEDED
+            elif (
+                e.error_code == databricks_pb2.ErrorCode.Name(databricks_pb2.REQUEST_LIMIT_EXCEEDED)
+                or e.get_http_status_code() not in _TRANSIENT_FAILURE_RESPONSE_CODES
```

Comment:
Nice catch!

Reply:
@balvisio Looks like an associated test (`TestRestStore.test_get_experiment_by_name[RestStore] `, https://github.com/mlflow/mlflow/runs/6990695520?check_suite_focus=true) is failing with

```
defresponse_fn(*args, **kwargs):
# pylint: disable=unused-argument
if kwargs.get("endpoint") == "/api/2.0/mlflow/experiments/get-by-name":
raise MlflowException(
>               "GetExperimentByName is not implemented", ENDPOINT_NOT_FOUND
            )
E           mlflow.exceptions.MlflowException: GetExperimentByName is not implemented
args       = ()
kwargs     = {'endpoint': '/api/2.0/mlflow/experiments/get-by-name', 'host_creds': <mlflow.utils.rest_utils.MlflowHostCreds object at 0x7f749f6a16d0>, 'method': 'GET', 'params': {'experiment_name': 'abc'}}
list_exp_response = <MagicMock id='140138615968208'>
```

Can we special case `ENDPOINT_NOT_FOUND` here as well & fall back to the `ListExperiments`-based implementation?

Reply:
@dbczumar yeah I saw the failure. What if the function is written like this:

```python
def get_experiment_by_name(self, experiment_name):
        try:
            req_body = message_to_json(GetExperimentByName(experiment_name=experiment_name))
            response_proto = self._call_endpoint(GetExperimentByName, req_body)
            return Experiment.from_proto(response_proto.experiment)
        except MlflowException as e:
            if e.error_code == databricks_pb2.ErrorCode.Name(
                databricks_pb2.RESOURCE_DOES_NOT_EXIST
            ):
                return None
            elif e.error_code == databricks_pb2.ErrorCode.Name(
                databricks_pb2.ENDPOINT_NOT_FOUND
            ):
                # Fall back to using ListExperiments-based implementation.
                for experiment in self.list_experiments(ViewType.ALL):
                    if experiment.name == experiment_name:
                        return experiment
                return None 
            raise e
```

I run the rest_store tests locally and they pass

Reply:
I think the implementation above is not fully correct though because we are populating the error code from the HTTP status but since 404 maps both to ENDPOINT_NOT_FOUND and RESOURCE_DOES_NOT_EXIST the function could be returning `None` when it actually needs to do the listing method. A solution could be to add :

```python
HTTP_STATUS_TO_ERROR_CODE = dict((v, k) for k, v in ERROR_CODE_TO_HTTP_STATUS.items())
HTTP_STATUS_TO_ERROR_CODE[404] = ErrorCode.Name(ENDPOINT_NOT_FOUND)
```

to error on the safe side

Reply:
Also a question I have is: why if the ENDPOINT_NOT_FOUND then `list_experiment` help? Doesn't this function also calls the same endpoint?

Reply:
@balvisio Thanks for calling this out! Can we update `mlflow.utils.rest_utils.get_error_code` to check the response *text* if the error code is 404 and, if the text contains `"ENDPOINT_NOT_FOUND"`, return the "endpoint not found" error code representation; otherwise, the "resource does not exist" error code representation should be returned.

Reply:
@dbczumar I think that removing `RESOURCE_DOES_NOT_EXIST` from the mapping is the safe option because we will be defaulting to `ENDPOINT_NOT_FOUND` which it will cause the code to try the `list_experiment()` path. 

From looking at the tests, I thought that the MLFlow server specifically returned a JSON struct containing: `RESOURCE_DOES_NOT_EXIST`: 

```
{
    "error_code": "RESOURCE_DOES_NOT_EXIST", 
    "message": "Node type not supported"
}
```

which is handled in this branch of `verify_rest_response`:

```
if _can_parse_as_json_object(response.text):
    raise RestException(json.loads(response.text))
else:
...
```

The reason I am reluctant to look at the text message is that if there is a proxy server between the MLflow server and the client (for example, nginx) the string "API not found" might or might be there depending on the proxy.

Let me know if you still want to go ahead with https://github.com/mlflow/mlflow/pull/6106#discussion_r904376705 and I can update the PR.

Reply:
@balvisio Apologies for the delay. I think that defaulting to `RESOURCE_DOES_NOT_EXIST` for 404 is a better option, since `ENDPOINT_NOT_FOUND` is a more specific status that is not commonly used throughout MLflow. Rather than checking for message text to parse `ENDPOINT_NOT_FOUND`, which I agree is brittle, could we instead rely on the fact that `ENDPOINT_NOT_FOUND` is parsed as a `RestException` in the `get_experiment_by_name()` exception handler?

If that approach doesn't work, another option is to un-indent the preexisting `list_experiments()` fallback logic. This way, we won't encounter the nested exception. I realize that we may end up falling back to `list_experiments()` for irrecoverable errors (e.g. auth errors) in this case, but that's consistent with preexisting behavior.

Reply:
@dbczumar thanks for the reply. I got the point of `ENDPOINT_NOT_FOUND` not being used frequently. 

However, by injecting `ENDPOINT_NOT_FOUND` in `rest_utils.py:191` (using the `get_error_code()` function) and then calling `list_experiments()` in the `ENDPOINT_NOT_FOUND` case (`rest_store.py:333`), we are effectively "falling back" to calling `list_experiments()` path which is what we want.
I also just noticed the implementation `get_experiment_by_name()` of the `DatabricksRestStore` (`rest_store.py:325`) uses the same logic.

Reply:
Right, we want to effectively fall back. It seems like, if the `ENDPOINT_NOT_FOUND` error is parsed as a `RestException`, which I think is the case, we can achieve this fallback logic without mapping the 404 status code to `ENDPOINT_NOT_FOUND`. I.e. tests should still pass and the behavior should still be correct if we change


```
# Removing 'RESOURCE_DOES_NOT_EXIST' from dict to disambiguate between RESOURCE_DOES_NOT_EXIST and
# ENDPOINT_NOT_FOUND. When 'get_error_code()' is used it is the ENDPOINT_NOT_FOUND case since when
# RESOURCE_DOES_NOT_EXIST the response content is (and parsed as) a JSON object.
HTTP_STATUS_TO_ERROR_CODE = dict(
    (v, k)
    for k, v in ERROR_CODE_TO_HTTP_STATUS.items()
    if k != ErrorCode.Name(RESOURCE_DOES_NOT_EXIST)
)
```

to 

```
HTTP_STATUS_TO_ERROR_CODE = dict(
    (v, k)
    for k, v in ERROR_CODE_TO_HTTP_STATUS.items()
)
HTTP_STATUS_TO_ERROR_CODE[404] = ErrorCode.Name(RESOURCE_DOES_NOT_EXIST)
```

Reply:
@dbczumar Great. Updated the PR. Not sure why in the last run some pipelines failed at the beginning; looked like some spurious error.